### PR TITLE
fix(cast) Create2 init code miscalculation

### PR DESCRIPTION
**Pull request by iFrostizz**
Following up on https://github.com/foundry-rs/foundry/issues/3510
Fixing the bug, and adding tests.

### CODE REVIEW

File: cli/src/cmd/cast/create2.rs [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+            if starts_with.is_some() || ends_with.is_some() {
+                eyre::bail!("Either use --matching or --starts/ends-with");
+            }
+
+            if matches.len() != 40 {
+                if !(matches.len() == 42 && matches.starts_with("0x")) {
+                    eyre::bail!("Please provide a 40 characters long sequence for matching");
+                }
+                matches = matches.strip_prefix("0x").unwrap().to_string();
+            }
```
Comment (mattsse) [inline cli/src/cmd/cast/create2.rs:95]: can just `trim_start_matches("0x")` and then do a len check with the `bail`

File: cli/src/cmd/cast/create2.rs [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+
+            hex::decode(&matches.replace('X', "0")).wrap_err("invalid matching hex provided")?;
+            regexs.push(matches.replace('X', "."));
```
Comment (mattsse) [inline cli/src/cmd/cast/create2.rs:98]: can we add a brief note why we do this?

File: cli/src/cmd/cast/create2.rs [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+
+        assert!(
+            regexs.iter().map(|p| p.len() - 1).sum::<usize>() <= 40,
+            "vanity patterns length exceeded. cannot be more than 40 characters",
+        );
+
+        let regex = RegexSetBuilder::new(regexs).case_insensitive(!case_sensitive).build()?;
+
+        let init_code_hash = if let Some(init_code_hash) = init_code_hash {
+            let mut a: [u8; 32] = Default::default();
```
Comment (mattsse) [inline cli/src/cmd/cast/create2.rs:122]:
```suggestion
            let mut a: [u8; 32] = [0; 32];
```

File: cli/src/cmd/cast/create2.rs [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+            let pad_width = suffix.len() + suffix.len() % 2;
+            hex::decode(format!("{:0>width$}", suffix, width = pad_width))
+                .wrap_err("invalid suffix hex provided")?;
+            regexs.push(format!(r"{}$", suffix));
+        }
+
+        assert!(
+            regexs.iter().map(|p| p.len() - 1).sum::<usize>() <= 40,
+            "vanity patterns length exceeded. cannot be more than 40 characters",
+        );
```
Comment (mattsse) [inline cli/src/cmd/cast/create2.rs:117]: is this a sanity `assert!` or something that depends on user input?

for former, can do a `debug_assert!` for latter, handle this as error.

File: cli/src/cmd/cast/create2.rs [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+                let addr = SimpleCast::to_checksum_address(&get_create2_address_from_hash(
+                    deployer,
+                    salt.clone(),
+                    init_code_hash,
+                ));
+
+                (salt, addr)
+            })
+            .find_any(move |(_, addr)| {
+                let addr = addr.to_string();
```
Comment (mattsse) [inline cli/src/cmd/cast/create2.rs:149]: is `addr` of type `Address` here?

because the `Display` impl (`.to_string`) truncates.

**Review by mattsse — CHANGES_REQUESTED** [review]
thanks,

like the directions, but have some suggestions.

needs a rebase and nightly formatting `cargo +nightly fmt`

File: cli/src/cmd/cast/create2.rs (hunk 4 above)
Comment (iFrostizz) [reply cli/src/cmd/cast/create2.rs]: `debug_assert!` seems also to `panic!`, I guess that I should use `eyre::bail!` here.

File: cli/src/cmd/cast/create2.rs (hunk 5 above)
Comment (iFrostizz) [reply cli/src/cmd/cast/create2.rs]: I need to convert it to `String` here for the regex matching, agree with you that it truncates but it works, was the `Display` recently changed ? Logging the address on the fly doesn't shows it truncated.

**Comment by iFrostizz** [comment]
Okay I guess that I will need some help on the rebase here....

**Comment by gakonst** [comment]
I would just `git reset --hard master` and `git cherry-pick <..>` the changes you made from there.

**Comment by iFrostizz** [comment]
Now should be gud :) Thanks

**Comment by iFrostizz** [comment]
Can I add a little bump @mattsse ? :smile:

**Review by mattsse — APPROVED** [review]
sorry for the wait, this one slipped through...

tysm, looks fine, love the tests.

just rebased, merging ones CI clears.

**Pull request merged**